app/src/main/java/com/example/photostomusic/fragments/RecommendationFragment.java

```diff
@@ -249,27 +254,53 @@ public void discarded(int mIndex, int direction) {
                                 case 0:
                                     // Song rejected, it is now removed from stack
                                     Log.d(TAG, "discarded: 0");
+                                    mp.stop();
+                                    mp.reset();
                                     break;
                                 case 1:
                                     // Song accepted, sending to Parse
                                     Log.d(TAG, "discarded: 1");
-                                    saveSong(songImage, songName, songArtist, songAlbum, songCode, genres, photo);
+                                    try {
+                                        saveSong(songs.getJSONObject(mIndex-1), genres, photo);
+                                    } catch (JSONException e) {
+                                        e.printStackTrace();
+                                    }
                                     break;
                                 case 2:
                                     // Song rejected, it is now removed from stack
                                     Log.d(TAG, "discarded: 2");
+                                    mp.stop();
+                                    mp.reset();
                                     break;
                                 case 3:
                                     // Accept, send to parse and move to history
                                     Log.d(TAG, "discarded: 3");
-                                    saveSong(songImage, songName, songArtist, songAlbum, songCode, genres, photo);
+                                    try {
+                                        saveSong(songs.getJSONObject(mIndex), genres, photo);
+                                    } catch (JSONException e) {
+                                        e.printStackTrace();
+                                    }
                                     break;
                             }
                         }
 
                         @Override
-                        // TODO: (STRETCH) Add preview of song with Spotify API
                         public void topCardTapped() {
+                            mp.stop();
+                            mp.reset();
```

Comment:
also maybe pull these two (stop, reset) into a function to reduce duplicate code

Reply:
Done :)